resources/assets/lib/beatmap-discussions/new-discussion.tsx

```diff
@@ -0,0 +1,517 @@
+  discussions: Discussion[];
+  timestamp: number | null;
+}
+
+interface Props {
+  autoFocus: boolean;
+  beatmapset: BeatmapsetExtendedJson & BeatmapsetWithDiscussionsJson;
+  currentBeatmap: BeatmapExtendedJson;
+  currentDiscussions: CurrentDiscussions;
+  currentUser: CurrentUserJson;
+  innerRef: React.RefObject<HTMLDivElement>;
+  mode: Mode;
+  pinned: boolean;
+  setPinned?: (flag: boolean) => void;
+  stickTo: React.RefObject<HTMLElement>;
+}
+
+@observer
+export class NewDiscussion extends React.Component<Props> {
+  static readonly defaultProps = {
+    pinned: false,
+  };
+
+  private readonly disposers = new Set<((() => void) | undefined)>();
+  private readonly handleKeyDown;
+  private readonly inputBox = React.createRef<HTMLTextAreaElement>();
+  @observable private message = this.storedMessage;
+  private nearbyDiscussionsCache: DiscussionsCache | null = null;
+  @observable private posting: string | null = null;
+  private postXhr: JQuery.jqXHR<BeatmapsetDiscussionPostStoreResponseJson> | null = null;
+  @observable private sticky = this.props.pinned;
+  @observable private timestampConfirmed = false;
+
+  private get canPost() {
+    return !this.props.currentUser.is_silenced
+      && (!this.props.beatmapset.discussion_locked
+        || canModeratePosts(this.props.currentUser))
+      && (this.props.currentBeatmap.deleted_at == null || this.props.mode === 'generalAll');
+  }
+
+  @computed
+  private get cssTop() {
+    if (!this.sticky || this.props.stickTo?.current == null) return;
+    return core.stickyHeader.headerHeight + this.props.stickTo.current.getBoundingClientRect().height;
+  }
+
+  private get isTimeline() {
+    return this.props.mode === 'timeline';
+  }
+
+  private get nearbyDiscussions() {
+    const timestamp = this.timestamp;
+    if (timestamp == null) return [];
+
+    if (this.nearbyDiscussionsCache == null || (this.nearbyDiscussionsCache.beatmap !== this.props.currentBeatmap || this.nearbyDiscussionsCache.timestamp !== this.timestamp)) {
+      this.nearbyDiscussionsCache = {
+        beatmap: this.props.currentBeatmap,
+        discussions: BeatmapDiscussionHelper.nearbyDiscussions(this.props.currentDiscussions.timelineAllUsers, timestamp),
+        timestamp: this.timestamp,
+      };
+    }
+
+    return this.nearbyDiscussionsCache?.discussions ?? [];
+  }
+
+  private get storageKey() {
+    return `beatmapset-discussion:store:${this.props.beatmapset.id}:message`;
+  }
+
+  private get storedMessage() {
+    return localStorage.getItem(this.storageKey) ?? '';
+  }
+
+  @computed
+  private get timestamp() {
+    return this.props.mode === 'timeline'
+      ? BeatmapDiscussionHelper.parseTimestamp(this.message)
+      : null;
+  }
+
+  private get validPost() {
+    if (!validMessageLength(this.message, this.isTimeline)) return false;
+
+    if (this.isTimeline) {
+      return this.timestamp != null && (this.nearbyDiscussions.length === 0 || this.timestampConfirmed);
+    }
```

Comment:
This is indeed keeping the current behavior but there's suggestion for different behavior for praise and note buttons (#5105)